### ISSUE TITLE
[7.7 cs] internal error in event code

### ISSUE THREAD

**Issue by 97jaz** [racket/racket#3229]
Here's a very simple program that reproduces the problem mentioned in #2726:
```racket
#lang racket/base

(define out1 (open-output-bytes))
(define out2 (open-output-bytes))
(define evt (replace-evt out1 (λ (_) out2)))
(sync (handle-evt evt (λ (out) (write-bytes #"hello world!" out))))
```

In Racket 7.7 cs, this fails with `internal error: choice event discovered after interrupt/retry callbacks`.

I'd be happy to look into this myself, but I could use a primer to the event code. The error originates [here](https://github.com/racket/racket/blob/8b8e3913a83fd61d1406a12bf5735b0e36d54c7b/racket/src/thread/sync.rkt#L443) inside `sync-poll`, but I don't yet grok why `choice-evt`s should all be disposed of at this point.

**Comment by 97jaz** [comment]
It looks like a lot of the published material on CML isn't freely available, but I did read [this paper](https://people.cs.uchicago.edu/~jhr/papers/1991/pldi-cml.pdf), in which there is nothing corresponding to `replace-evt`. Just using the combinators available in that paper, we could do:
```racket
(define evt
  (guard-evt (λ ()
               (sync out1)
               out2)))
```
And that causes no problems with 7.7 cs. Obviously `replace-evt` isn't defined in terms of `guard-evt` this way (in fact, `guard-evt` appears to be implemented at a higher level), so I'm curious: what is the advantage of `replace-evt`?

**Comment by mflatt** [comment]
`replace-evt` is a Racket addition to CML.

That `guard-evt` pattern isn't what you want, because a `choice-evt` combinining that event and a different one cannot select the other one if it becomes ready before `out1`.

**Comment by mflatt** [comment]
I'll have to look more, but an initial guess is that `replace-evt` needs to remove interrupt and retry wrappers when it replaces an event.

**Comment by mflatt** [comment]
Thanks for the small example! Here's an even simpler one:
```
(define out1 always-evt)
(define out2 (choice-evt out1 always-evt))
(define evt (replace-evt out1 (λ (_) out2)))
(sync evt)
```
The problem is that a `replace-evt` installs setup and teardown actions for a `sync` to propagate them to nested evts within in the first evt. (A setup action might put a thread in a semaphore's queue, for example, and the accompanyng teardown removes it when the thread is no longer interested.) Once that first evt is ready and selected, the setup and teardown actions for a `replace-evt` are no longer relevant, so they should be removed as the second evt in `replace-evt` replavces the overall `replace-evt`.

Possibly, the right general rule is that setup and teardown actions should be discarded when an evt's polling procedure returns a replacement event. The  replacement can happen in all sorts of situations, such as `guard-evt`, but maybe `replace-evt` is the only case where the evt may supply setup and teardown actions on the way to producing a replacement.

**Comment by mflatt** [comment]
My analysis left out the abandon callback, which is similar to setup and teardown, but is used when an event is not selected. A `nack-guard-evt` wrapped about a `replace-evt` creates a abandon callback that needs to be preserved, so it's more complicated than discarding all of the callbacks on a replace.

**Comment by mflatt** [comment]
Looking yet again, I see that there's no complaint when abandon callbacks are present, and they are already handled correctly, so nevermind that part.

**Comment by 97jaz** [comment]
> Possibly, the right general rule is that setup and teardown actions should be discarded when an evt's polling procedure returns a replacement event. The replacement can happen in all sorts of situations, such as guard-evt, but maybe replace-evt is the only case where the evt may supply setup and teardown actions on the way to producing a replacement.

Replacement happens a lot, right? A poller returns two values, and if the first is `#f`, then the second will be an event that replaces the current one, IIUC. But from what I can tell, the wrappers installed when `sync-poll` encounters a `control-state-evt` (which I _think_ are the ones relevant to this case...) never get removed. 

It's unclear to me under what conditions a `syncer` gets reused, since in some places, at least, it seems that new ones are constructed (by `evts-syncers`, for example).

**Comment by mflatt** [comment]
Yes, replacement happens often, and the potential for triggering the "general rule" too much leads me to a different solution: allow `'reset` in a `control-state-evt` to clear the current handler.

**Comment by 97jaz** [comment]
Thanks Matthew!
